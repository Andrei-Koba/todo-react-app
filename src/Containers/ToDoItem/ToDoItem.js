import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { PropTypes } from 'prop-types';
import styles from './ToDoItem.module.scss';
import ToDoCheckbox from '../../Components/ToDoCheckbox/ToDoCheckbox';
import DeleteButton from '../../Components/DeleteButton/DeleteButton';
import ToDoDescription from '../../Components/ToDoDescription/ToDoDescription';
import { changeItemStatus, removeItem } from '../../Redux/StateSlices/todoSlice';

export default function ToDoItem({ item }) {
  const [mouseOn, setMouseOn] = useState(false);
  const dispatch = useDispatch();

  const handleCheckboxChange = () => {
    dispatch(changeItemStatus(item.id));
  };

  const handleDeleteClick = () => {
    dispatch(removeItem(item.id));
  };

  const handleMouseEnter = () => setMouseOn(true);
  const handleMouseLeave = () => setMouseOn(false);

  return (
    <li
      key={item.id}
      className={`${styles.todoItem} ${mouseOn ? styles.mouseOn : ''}`}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <label htmlFor={item.id} className={styles.checkbox}>
        <ToDoCheckbox
          id={item.id}
          isChecked={item.isCompleted}
          handleChange={handleCheckboxChange}
        />
        <ToDoDescription isCompleted={item.isCompleted} text={item.text} />
      </label>
      {mouseOn && <DeleteButton handleClick={handleDeleteClick} />}
    </li>
  );
}

ToDoItem.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    isCompleted: PropTypes.bool.isRequired,
  }).isRequired,
};
