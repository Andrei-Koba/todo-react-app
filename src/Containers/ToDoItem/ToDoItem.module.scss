.todoItem {
    display: flex;
    justify-content: space-between;
    padding: 15px;
    font-weight: 400;
    border-bottom: 1px solid lightgray;
}

.checkbox {
    display: flex;
    cursor: pointer;

    input[type="checkbox"] {
        display: none;
    }
}

.mouseOn {
    box-shadow: 1px 1px 50px lightgray;
}