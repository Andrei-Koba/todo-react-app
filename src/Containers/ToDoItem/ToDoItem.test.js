import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render, fireEvent } from '@testing-library/react';
import ToDoItem from './ToDoItem';

const item = {
  id: 'testId',
  text: 'testText',
  isCompleted: false,
};
const mockStore = configureStore([]);

describe('ToDoItem', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({});
    component = render(
      <Provider store={store}>
        <ToDoItem item={item} />
      </Provider>,
    );
  });

  it.each([true, false])('renders correctly, mouseOn="%p"', (mouseOn) => {
    if (mouseOn) {
      const listItem = component.getByRole('listitem');
      fireEvent.mouseEnter(listItem);
    }
    expect(component).toMatchSnapshot();
  });

  it('add class on mouseEnter', async () => {
    const listItem = component.getByRole('listitem');
    fireEvent.mouseEnter(listItem);
    expect(listItem.classList.contains('mouseOn')).toBe(true);
    fireEvent.mouseLeave(listItem);
    expect(listItem.classList.contains('mouseOn')).toBe(false);
  });
});
