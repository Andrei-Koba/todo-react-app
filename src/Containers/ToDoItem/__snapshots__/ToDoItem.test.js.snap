// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoItem renders correctly, mouseOn="false" 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="todoItem "
      >
        <label
          class="checkbox"
          for="testId"
        >
          <div>
            <div
              class="styledCheckbox unchecked"
            />
            <input
              id="testId"
              type="checkbox"
            />
          </div>
          <span
            class="itemText undefined"
          >
            testText
          </span>
        </label>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="todoItem "
    >
      <label
        class="checkbox"
        for="testId"
      >
        <div>
          <div
            class="styledCheckbox unchecked"
          />
          <input
            id="testId"
            type="checkbox"
          />
        </div>
        <span
          class="itemText undefined"
        >
          testText
        </span>
      </label>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ToDoItem renders correctly, mouseOn="true" 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="todoItem mouseOn"
      >
        <label
          class="checkbox"
          for="testId"
        >
          <div>
            <div
              class="styledCheckbox unchecked"
            />
            <input
              id="testId"
              type="checkbox"
            />
          </div>
          <span
            class="itemText undefined"
          >
            testText
          </span>
        </label>
        <input
          class="deleteButton"
          type="button"
        />
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="todoItem mouseOn"
    >
      <label
        class="checkbox"
        for="testId"
      >
        <div>
          <div
            class="styledCheckbox unchecked"
          />
          <input
            id="testId"
            type="checkbox"
          />
        </div>
        <span
          class="itemText undefined"
        >
          testText
        </span>
      </label>
      <input
        class="deleteButton"
        type="button"
      />
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
